### PR TITLE
#392: RqMethod validates the HTTP request method line.

### PR DESCRIPTION
**Pull request by bdragan**
#392: RqMethod validates the HTTP request method line.

### CODE REVIEW

**Comment by davvd** [comment]
@bdragan Thanks, let me find someone who can review this pull request

**Comment by davvd** [comment]
@dmzaytsev it's yours, please review

File: src/main/java/org/takes/rq/RqMethod.java [hunk 1]
```diff
@@ -94,6 +96,13 @@
     final class Base extends RqWrap implements RqMethod {
 
         /**
+         * HTTP method line pattern.
+         */
+        private static final Pattern PATTERN = Pattern.compile(
+            "(GET|POST|PUT|DELETE|HEAD|OPTIONS|PATCH) [^ ]+( [^ ]+){0,1}"
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqMethod.java]: @bdragan what about extension-method that the ticket reporter mentioned?

File: src/test/java/org/takes/rq/RqMethodTest.java [hunk 2]
```diff
@@ -47,4 +49,54 @@ public void returnsMethod() throws IOException {
...
+     * RqMethod can fail when HTTP method line has any extra undefined
+     * elements.
+     * @throws IOException If some problem inside
+     */
+    @Test(expected = InvalidHTTPMethodLineException.class)
+    public void failsOnExtraLineElement() throws IOException {
+        new RqMethod.Base(
+            new RqSimple(Arrays.asList("GET / HTTP/1.1 abc"), null)
+        ).method();
+    }
```
Comment (dmzaytsev) [inline src/test/java/org/takes/rq/RqMethodTest.java]: @bdragan lets add a new test for the extension-method

File: src/test/java/org/takes/rq/RqMethodTest.java [hunk 3]
```diff
@@ -47,4 +49,54 @@ public void returnsMethod() throws IOException {
...
+     * RqMethod supports all HTTP methods.
+     * @throws IOException If some problem inside
+     */
+    @Test
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    public void supportsAllMethods() throws IOException {
+        for (final String method
+            : Arrays.asList(
+                RqMethod.DELETE, RqMethod.GET, RqMethod.HEAD, RqMethod.OPTIONS,
+                RqMethod.PATCH, RqMethod.POST, RqMethod.PUT
```
Comment (dmzaytsev) [inline src/test/java/org/takes/rq/RqMethodTest.java]: @bdragan there is method `TRACE` in [RFC](http://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html)

File: src/main/java/org/takes/rq/RqMethod.java (hunk 1 above)
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqMethod.java]: @bdragan there is only one space allows after uri (http://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html)
```
Request-Line   = Method SP Request-URI SP HTTP-Version CRLF
```
but the method line like the string below will be valid
```
GET http://www.w3.org/pub/WWW/TheProject.html                               HTTP/1.1
```

**Comment by dmzaytsev** [comment]
@bdragan a few comments above

File: src/main/java/org/takes/rq/RqMethod.java (hunk 1 above)
Comment (bdragan) [inline src/main/java/org/takes/rq/RqMethod.java]: @dmzaytsev I allow only one space in the regex part after the URI group of characters: ( [^ ]+)
All of the other characters must be non-space ones: [^ ]

**Comment by bdragan** [comment]
@dmzaytsev Thanks, please see the new changes.

**Comment by dmzaytsev** [comment]
@bdragan thanks

**Comment by dmzaytsev** [comment]
@rultor merge

File: src/main/java/org/takes/rq/RqMethod.java [hunk 4]
```diff
@@ -104,8 +131,34 @@ public Base(final Request req) {
         @Override
         public String method() throws IOException {
             final String line = this.head().iterator().next();
-            final String[] parts = line.split(" ", 2);
-            return parts[0].toUpperCase(Locale.ENGLISH);
+            final Matcher matcher = PATTERN.matcher(line);
+            if (!matcher.matches()) {
+                throw new InvalidHTTPMethodLineException(line);
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqMethod.java]: @bdragan what is the point? why can't we just throw `IOException`?

**Comment by yegor256** [comment]
@bdragan this exception class is redundant, just `IOException` is enough, see http://www.yegor256.com/2015/07/28/checked-vs-unchecked-exceptions.html

**Comment by bdragan** [comment]
@yegor256 Thanks; fixed.

**Comment by yegor256** [comment]
@bdragan I understand why you're introducing this supplementary method - to avoid code duplication. But it's a bad idea in this case. Don't be afraid of code duplication, if your code is doing different things. First, it throws an exception if something is wrong. Then, it throws if something **else** is wrong. There is **nothing** in common between these two situations. Get rid of that static method.

**Comment by bdragan** [comment]
@yegor256 True, in this case it is not needed; even the message should be a bit different. Thanks; fixed.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@ypshenychka please, help us to evaluate quality here, as per [par.24](http://at.teamed.io/policy.html#24)

**Comment by davvd** [comment]
@rultor deploy now pls

**Comment by ypshenychka** [comment]
@davvd Quality is good.

**Comment by davvd** [comment]
> @davvd Quality is good.
> 

@ypshenychka thank you, good quality is always good ;)

**Comment by davvd** [comment]
@dmzaytsev added 10 mins to @ypshenychka, for QA review, payment code is `71781473`; Much obliged! I have added **22 mins** to your account in payment "71781502", 121 hours and 23 mins spent; you're getting extra minutes here (c=7); added +22 to your rating, now it is equal to [+904](http://www.netbout.com/b/35166?open=rating)

**Pull request merged**